Mathlib/Combinatorics/Additive/Convolution.lean

```diff
@@ -0,0 +1,95 @@
+Authors: Yaël Dillies, Strahinja Gvozdić, Bhavik Mehta
+-/
+import Mathlib.Algebra.Group.Action.Pointwise.Finset
+
+/-!
+# Convolution
+
+This file defines convolution of finite subsets `A` and `B` of group `G` as the map `A ⋆ B : G → ℕ`
+that maps `x ∈ G` to the number of distinct representations of `x` in the form `x = ab` for
+`a ∈ A`, `b ∈ B`. It is shown how convolution behaves under the change of order of `A` and `B`, as
+well as under the left and right actions on `A`, `B`, and the function argument.
+-/
+
+open MulOpposite MulAction
+open scoped Pointwise RightActions
+
+namespace Finset
+variable {G : Type*} [Group G] [DecidableEq G] {A B : Finset G} {a b s x y : G}
+
+/-- Given finite subsets `A` and `B` of a group `G`, convolution of `A` and `B` is a map `G → ℕ`
+that maps `x ∈ G` to the number of distinct representations of `x` in the form `x = ab`, where
+`a ∈ A`, `b ∈ B`. -/
+@[to_additive addConvolution /-- Given finite subsets `A` and `B` of an additive group `G`,
+convolution of `A` and `B` is a map `G → ℕ` that maps `x ∈ G` to the number of distinct
+representations of `x` in the form `x = a + b`, where `a ∈ A`, `b ∈ B`. -/]
+def convolution (A B : Finset G) : G → ℕ := fun x => #{ab ∈ A ×ˢ B | ab.1 * ab.2 = x}
+
+@[to_additive]
+lemma card_smul_inter_smul (A B : Finset G) (x y : G) :
+    #((x • A) ∩ (y • B)) = A.convolution B⁻¹ (x⁻¹ * y) :=
+  card_nbij' (fun z ↦ (x⁻¹ * z, z⁻¹ * y)) (fun ab' ↦ x • ab'.1)
+    (by simp +contextual [Set.MapsTo, Set.mem_smul_set_iff_inv_smul_mem, mul_assoc])
+    (by simp +contextual [Set.MapsTo, Set.mem_smul_set_iff_inv_smul_mem]
+        simp +contextual [← eq_mul_inv_iff_mul_eq, mul_assoc])
+    (by simp [Set.LeftInvOn])
+    (by simp +contextual [Set.LeftInvOn, ← eq_mul_inv_iff_mul_eq, mul_assoc])
+
+@[to_additive (attr := simp)]
+lemma card_inter_smul (A B : Finset G) (x : G) : #(A ∩ (x • B)) = A.convolution B⁻¹ x := by
+  simpa using card_smul_inter_smul _ _ 1 x
+
+@[to_additive (attr := simp)]
+lemma card_smul_inter (A B : Finset G) (x : G) : #((x • A) ∩ B) = A.convolution B⁻¹ x⁻¹ := by
+  simpa using card_smul_inter_smul _ _ x 1
+
+@[to_additive addConvolution_pos]
+lemma convolution_pos : 0 < A.convolution B x ↔ x ∈ A * B := by
+  constructor
+  · intro (hx : 0 < #{ab ∈ _ | _})
+    obtain ⟨ab, hab⟩ := card_pos.mp hx
+    obtain ⟨⟨ha, hb⟩, hab⟩ := by simpa only [mem_filter, mem_product] using hab
+    simpa only [← hab] using mul_mem_mul ha hb
+  · intro hx
+    obtain ⟨a, ha, b, hb, h⟩ := mem_mul.mp hx
+    exact card_pos.mpr ⟨(a, b), by simp [ha, hb, h, mem_filter, mem_product]⟩
+
+@[to_additive addConvolution_le_card_left]
+lemma convolution_le_card_left (A B : Finset G) (x : G) : A.convolution B x ≤ #A := by
+  rw [← inv_inv B, ← card_inter_smul]
+  exact card_le_card inter_subset_left
+
+@[to_additive addConvolution_le_card_right]
+lemma convolution_le_card_right (A B : Finset G) (x : G) : A.convolution B x ≤ #B := by
```

Comment:
I would make all arguments implicit in these two, since they can't be used for rewriting
```suggestion
lemma convolution_le_card_left : A.convolution B x ≤ #A := by
  rw [← inv_inv B, ← card_inter_smul]
  exact card_le_card inter_subset_left

@[to_additive addConvolution_le_card_right]
lemma convolution_le_card_right : A.convolution B x ≤ #B := by
```